core-bundle/src/EventListener/DataContainer/UndoListener.php

```diff
@@ -0,0 +1,347 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\ArrayUtil;
+use Contao\Backend;
+use Contao\Config;
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\ServiceAnnotation\Callback;
+use Contao\DataContainer;
+use Contao\Date;
+use Contao\Image;
+use Contao\StringUtil;
+use Contao\System;
+use Contao\UserModel;
+use Doctrine\DBAL\Connection;
+
+class UndoListener
+{
+    private Connection $connection;
+    private ContaoFramework $framework;
+
+    /**
+     * @var Controller
+     */
+    private $controller;
+
+    /**
+     * @var Backend
+     */
+    private $backend;
+
+    /**
+     * @var Image
+     */
+    private $image;
+
+    public function __construct(ContaoFramework $framework, Connection $connection)
+    {
+        $this->framework = $framework;
+        $this->connection = $connection;
+
+        /** @var Controller $controller */
+        $controller = $framework->getAdapter(Controller::class);
+        $this->controller = $controller;
+
+        /** @var Backend $backend */
+        $backend = $framework->getAdapter(Backend::class);
+        $this->backend = $backend;
+
+        /** @var Image $image */
+        $image = $this->framework->getAdapter(Image::class);
+        $this->image = $image;
```

Comment:
The framework will need to be initialized at this point. It's probably better to move these calls out of the constructor + call `$framework->initialize()` before the first use.